### PR TITLE
fix(k8s): virtual probes for sidecar initContainer ports also exposed by a Service

### PR DESCRIPTION
**Pull request by michaelbeaumont**
### Checklist prior to review

- [x] [Link to relevant issue][1] as well as docs and UI issues --
- [x] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [x] Tests (Unit test, E2E tests, manual test on universal and k8s) --
  - Don't forget `ci/` labels to run additional/fewer tests
- [x] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [x] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? ([this](https://github.com/kumahq/kuma/actions/workflows/auto-backport.yaml) GH action will add "backport" label based on these [file globs](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L6), if you want to prevent it from adding the "backport" label use [no-backport-autolabel](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L8) label) --

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

File: pkg/plugins/runtime/k8s/controllers/probes.go [hunk 1]
```diff
@@ -30,6 +30,37 @@ func ProbesFor(pod *kube_core.Pod) (*mesh_proto.Dataplane_Probes, error) {
...
 	}
+	var kumaSidecarSeen bool
+	for _, c := range pod.Spec.InitContainers {
+		if c.Name == util.KumaSidecarContainerName {
+			kumaSidecarSeen = true
+			continue
+		}
+		if !kumaSidecarSeen || c.RestartPolicy == nil || *c.RestartPolicy != kube_core.ContainerRestartPolicyAlways {
+			continue
+		}
```
Comment (jijiechen) [inline pkg/plugins/runtime/k8s/controllers/probes.go:41]: Here, I see regardless `kumaSidecarSeen` equals  `true` or not, the execution flow will match at least  one `continue`. Will the other statements work as expected?

File: pkg/plugins/runtime/k8s/webhooks/injector/probes.go [hunk 2]
```diff
@@ -29,6 +29,37 @@ func (i *KumaInjector) overrideHTTPProbes(pod *kube_core.Pod) error {
 		return err
 	}
 
+	var kumaSidecarSeen bool
+	for _, c := range pod.Spec.InitContainers {
+		if c.Name == util.KumaSidecarContainerName {
+			kumaSidecarSeen = true
+			continue
+		}
+		if !kumaSidecarSeen || c.RestartPolicy == nil || *c.RestartPolicy != kube_core.ContainerRestartPolicyAlways {
```
Comment (jijiechen) [inline pkg/plugins/runtime/k8s/webhooks/injector/probes.go:38]: Same question as previous

File: pkg/plugins/runtime/k8s/controllers/probes.go (hunk 1 above)
Comment (michaelbeaumont) [reply pkg/plugins/runtime/k8s/controllers/probes.go:41]: Once we have `kumaSidecarSeen` then we won't take either continue if we have a non Kuma sidecar and `RestartPolicy == Always`

Comment (jijiechen) [reply pkg/plugins/runtime/k8s/controllers/probes.go:41]: OK, understood, you are correct. The logic is correct here, but not intuitive to understand

Comment (jijiechen) [reply pkg/plugins/runtime/k8s/controllers/probes.go:41]: Can it be splitted into two parts?   The first part is responsible for finding our native sidecar, and there should be only one or none match; The second part is responsible for overriding the probes, if a match is found

Comment (michaelbeaumont) [reply pkg/plugins/runtime/k8s/controllers/probes.go:41]: I'm not 100% sure what you mean by finding the sidecar but I reduced duplication.

Comment (jijiechen) [reply pkg/plugins/runtime/k8s/controllers/probes.go:40]: I basically mean splitting the whole block into two stages and we can also convert the logic assertions to the opposite:

```go
var kumaSidecarSeen bool
	var initContainersToOverride []kube_core.Container
	for _, c := range pod.Spec.InitContainers {
		if c.Name == util.KumaSidecarContainerName {
			kumaSidecarSeen = true
			continue
		}

		isNativeSidecar := c.RestartPolicy != nil && *c.RestartPolicy == kube_core.ContainerRestartPolicyAlways
		if kumaSidecarSeen && isNativeSidecar {
			initContainersToOverride = append(initContainersToOverride, c)
		}
	}

	for _, c := range initContainersToOverride {
		if c.LivenessProbe != nil && c.LivenessProbe.HTTPGet != nil {
			if endpoint, err := probeFor(c.LivenessProbe, port); err != nil {
				return nil, err
			} else {
				dpProbes.Endpoints = append(dpProbes.Endpoints, endpoint)
			}
		}
		if c.ReadinessProbe != nil && c.ReadinessProbe.HTTPGet != nil {
			if endpoint, err := probeFor(c.ReadinessProbe, port); err != nil {
				return nil, err
			} else {
				dpProbes.Endpoints = append(dpProbes.Endpoints, endpoint)
			}
		}
		if c.StartupProbe != nil && c.StartupProbe.HTTPGet != nil {
			if endpoint, err := probeFor(c.StartupProbe, port); err != nil {
				return nil, err
			} else {
				dpProbes.Endpoints = append(dpProbes.Endpoints, endpoint)
			}
		}
	}
```

**Review by jijiechen — APPROVED** [review]

File: pkg/plugins/runtime/k8s/controllers/probes.go (hunk 1 above)
Comment (michaelbeaumont) [reply pkg/plugins/runtime/k8s/controllers/probes.go:40]: OK I did the splitting, I can change the logic assertions, I agree it's easier to read when it's not negated

**Review by jakubdyszkiewicz — APPROVED** [review]
make sense to me

**Pull request merged**